### PR TITLE
Fix background schema updates failing over a large upgrade gap

### PR DESCRIPTION
**Pull request by H-Shay**
Prior to this, if Synapse was upgraded from a sufficiently old version foreground migrations touching the `stream_ordering` column of the events table would apply before a background job also touching the same column was finished, resulting in the background job failing. Fixes https://github.com/matrix-org/synapse/issues/15677.

This work supercedes #15724, which turned out to be unworkable.

### CODE REVIEW

File: synapse/storage/schema/main/delta/78/05_mitigate_stream_ordering_update_race.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            ALTER TABLE local_current_membership ADD CONSTRAINT event_stream_ordering_fkey
+            FOREIGN KEY (event_stream_ordering) REFERENCES events(stream_ordering2) NOT VALID;
+            """
+            cur.execute(add_lcm_sql)
+
+            add_rm_sql = """
+            ALTER TABLE room_memberships ADD CONSTRAINT event_stream_ordering_fkey
+            FOREIGN KEY (event_stream_ordering) REFERENCES events(stream_ordering2) NOT VALID;
+            """
+            cur.execute(add_rm_sql)
```
Comment (erikjohnston) [inline synapse/storage/schema/main/delta/78/05_mitigate_stream_ordering_update_race.py:70]: Hmm, I'm not seeing anywhere in the code where we actually validate these constraits?

Comment (H-Shay) [reply synapse/storage/schema/main/delta/78/05_mitigate_stream_ordering_update_race.py:70]: I don't think that has happened yet. These columns/constraints were introduced in https://github.com/matrix-org/synapse/pull/15356. On the discussion of that PR there is mention of validating them in the future, after a background job migrates historical data from the `stream_ordering` column of `events` to the new stream ordering columns added, but I don't see anywhere in the deltas after this that the background job was added and run or that the constraints were validated.

Comment (erikjohnston) [reply synapse/storage/schema/main/delta/78/05_mitigate_stream_ordering_update_race.py:70]: Ah, I see, fair!

**Review by erikjohnston — APPROVED** [review]

**Comment by H-Shay** [comment]
Merged #15887 into develop, thank you for the review.

**Pull request merged**